# NOTE: Please make a copy of this notebook before you start changing anything.

## Loading Data

We will fetch the first two chapters of Jane Austen's Pride and Prejudice from [Project Gutenberg](https://www.gutenberg.org/ebooks/1342)


### Function to fetch data


In [22]:
def fetch_text(raw_url):
  """
  Fetch and cache text content from a URL.

  Downloads text from the given URL and saves it locally so that
  repeated calls use the cached version instead of re-downloading.
  Prints a success or failure message depending on the outcome.

  Parameters:
    raw_url (str): The URL of the raw text file to fetch.

  Returns:
    str: The full text content from the URL, or an empty string
         if the request fails.
  """
  import requests
  ...
  import requests
  from pathlib import Path
  import hashlib

  CACHE_DIR = Path("cs_110_content/text_cache")
  CACHE_DIR.mkdir(parents=True, exist_ok=True)

  def _url_to_filename(url):
    url_hash = hashlib.sha1(url.encode("utf-8")).hexdigest()[:12]
    return CACHE_DIR / f"{url_hash}.txt"

  cache_path = _url_to_filename(raw_url)

  SUCCESS_MSG = "✅ Text fetched."
  FAILURE_MSG = "❌ Failed to fetch text."
  try:
    if not cache_path.exists():
      response = requests.get(raw_url, timeout=10)
      response.raise_for_status()
      text_data = response.text
      cache_path.write_text(text_data, encoding="utf-8")
    print(SUCCESS_MSG)
    return cache_path.read_text(encoding="utf-8")

  except Exception as e:
    print(FAILURE_MSG)
    print(f"Error: {e}")
    return ""

### Save the text in a variable

In [18]:
PRIDE_PREJUDICE_URL = "https://www.gutenberg.org/cache/epub/1342/pg1342.txt"

pride_prejudice_text = fetch_text(PRIDE_PREJUDICE_URL)

✅ Text fetched.


## Statistics about the data

In [23]:
def print_text_stats(text):
  """
  Print basic statistics about a text string.

  Counts and displays the number of characters, lines, and words
  in the provided text.

  Parameters:
    text (str): The input text to analyze.

  Returns:
    None
  """

  num_chars = len(text)

  lines = text.splitlines()
  num_lines = len(lines)

  num_words = 0
  for line in lines:
    words_in_line = line.split()
    num_words_in_line = len(words_in_line)
    num_words += num_words_in_line

  print(f"Number of characters: {num_chars}")
  print(f"Number of lines: {num_lines}")
  print(f"Number of words: {num_words}")

print_text_stats(pride_prejudice_text)

Number of characters: 748168
Number of lines: 14915
Number of words: 130415


In [30]:
def get_word_counts(text):
  """
  Count the frequency of each word in a text string.

  Splits the text into lines and words, converts each word to
  lowercase, and returns a dictionary mapping each word to its
  frequency. No punctuation removal or stop word filtering is applied.

  Parameters:
    text (str): The input text to process.

  Returns:
    dict: A dictionary where keys are lowercase words (str) and
          values are their frequency counts (int).
  """
  word_counts = {}
  lines = text.splitlines()
  word_counts = {}
  lines = text.splitlines()
  for line in lines:
    words = line.split()
    for word in words:
      word = word.lower()
      if word in word_counts:
        word_counts[word] += 1
      else:
        word_counts[word] = 1
  return word_counts

word_counts = get_word_counts(pride_prejudice_text)
print(word_counts)

{'\ufeffthe': 1, 'project': 83, 'gutenberg': 62, 'ebook': 8, 'of': 3917, 'pride': 33, 'and': 3534, 'prejudice': 11, 'this': 425, 'is': 873, 'for': 1059, 'the': 4800, 'use': 26, 'anyone': 20, 'anywhere': 3, 'in': 2002, 'united': 21, 'states': 11, 'most': 200, 'other': 184, 'parts': 6, 'world': 30, 'at': 816, 'no': 460, 'cost': 6, 'with': 1115, 'almost': 66, 'restrictions': 3, 'whatsoever.': 2, 'you': 1135, 'may': 199, 'copy': 8, 'it,': 119, 'give': 122, 'it': 1197, 'away': 63, 'or': 388, 're-use': 2, 'under': 38, 'terms': 38, 'license': 12, 'included': 4, 'online': 5, 'www.gutenberg.org.': 4, 'if': 329, 'are': 359, 'not': 1452, 'located': 7, 'states,': 4, 'will': 403, 'have': 867, 'to': 4325, 'check': 8, 'laws': 9, 'country': 25, 'where': 93, 'before': 170, 'using': 7, 'ebook.': 2, 'title:': 1, 'author:': 1, 'jane': 154, 'austen': 17, 'release': 2, 'date:': 1, 'june': 2, '1,': 1, '1998': 1, '[ebook': 1, '#1342]': 1, 'recently': 1, 'updated:': 1, 'february': 2, '10,': 1, '2026': 1, 'lang

# Lab Exercise

Do the following:
* Make a new Code Cell below and use Gemini to create a new function called `print_top_10_frequent_words` that will call the above get_word_counts() and print only the top 10 frequent words.
* Make a new Text Cell after that and write a one-two sentence summary of the code. <br>**HINT:** Use Gemini's "Explain Code" and prompt it to give you a good summary.
* Once you are done, download this notebook by clicking on the File Menu (below the name of the file), Download > Download .ipynb and submit it.

In [34]:
# prompt: Make a new Code Cell below and use Gemini to create a new function called print_top_10_frequent_words that will call the above get_word_counts() and print only the top 10 frequent words.

import operator

def print_top_10_frequent_words(text):
  """
  Print the 10 most frequently occurring words in a text string.

  Calls get_word_counts() to build a word frequency dictionary,
  sorts the results in descending order, and prints the top 10
  words along with their counts.

  Parameters:
    text (str): The input text to analyze.

  Returns:
    None
  """
  word_counts = get_word_counts(text)
  sorted_word_counts = dict(sorted(word_counts.items(), key=operator.itemgetter(1), reverse=True))
  top_10_words = list(sorted_word_counts.items())[:10]
  for word, count in top_10_words:
    print(f"{word}: {count}")

print_top_10_frequent_words(pride_prejudice_text)


the: 4800
to: 4325
of: 3917
and: 3534
a: 2060
in: 2002
her: 1988
was: 1819
i: 1764
she: 1645


In [ ]:
def print_top_10_frequent_words(text):
    """
    Print the 10 most frequently occurring words in a text string.

    Calls get_word_counts() to build a word frequency dictionary,
    sorts the results in descending order, and prints the top 10
    words along with their counts.

    Parameters:
        text (str): The input text to analyze.

    Returns:
        None
    """

In [ ]:
#!pip install spacy

In [ ]:
import spacy

In [ ]:
nlp = spacy.load('en_core_web_sm')

In [36]:
def word_tokenization_normalization(text):
  """
  Tokenize and normalize text using spaCy NLP preprocessing.

  Converts text to lowercase, processes it with a spaCy language
  model, and filters out stop words, punctuation, numbers, and
  tokens with two or fewer characters. Returns the lemmatized
  form of each remaining token.

  Parameters:
    text (str): The raw input text to process.

  Returns:
    list: A list of lemmatized, normalized word strings.
  """
  text = text.lower() # lowercase
  doc = nlp(text)     # loading text into model

  words_normalized = []
  for word in doc:
    if word.text != '\n' \
    and not word.is_stop \
    and not word.is_punct \
    and not word.like_num \
    and len(word.text.strip()) > 2:
      word_lemmatized = str(word.lemma_)
      words_normalized.append(word_lemmatized)

  return words_normalized

In [ ]:
doc_tokenized = word_tokenization_normalization(pride_prejudice_text)

In [ ]:
print(doc_tokenized)

In [38]:
# prompt: write a function that takes in a list of words and returns a word count dictionary

def word_count(word_list):
  """
  Count the frequency of each word in a list of words.

  Iterates through the provided list, converts each word to
  lowercase, and builds a dictionary mapping each unique word
  to the number of times it appears.

  Parameters:
    word_list (list): A list of word strings to count.

  Returns:
    dict: A dictionary where keys are lowercase words (str) and
          values are their frequency counts (int).
  """
  word_counts = {}
  for word in word_list:
    word = word.lower()
    if word in word_counts:
      word_counts[word] += 1
    else:
      word_counts[word] = 1
  return word_counts

In [ ]:
new_counts = word_count(doc_tokenized)
print(new_counts)

In [39]:
# prompt: write a function that takes in a word count dictionary and prints the top 15 words with the most counts

import operator

def print_top_15_frequent_words(word_counts):
  """
  Print the 15 most frequently occurring words from a word count dictionary.

  Sorts the dictionary by count in descending order and prints
  the top 15 words along with their counts.

  Parameters:
    word_counts (dict): A dictionary mapping words (str) to
                        their frequency counts (int).

  Returns:
    None
  """
  sorted_word_counts = dict(sorted(word_counts.items(), key=operator.itemgetter(1), reverse=True))
  top_15_words = list(sorted_word_counts.items())[:15]
  for word, count in top_15_words:
    print(f"{word}: {count}")

In [40]:
print_top_15_frequent_words(new_counts)

elizabeth: 639
darcy: 428
say: 426
know: 394
mrs: 354
bennet: 348
think: 346
miss: 319
bingley: 315
jane: 303
sister: 292
lady: 280
come: 268
good: 248
see: 245


The basic get_word_counts() function returns raw, unfiltered word frequencies, so its top 10 results are dominated by common function words like "the" (4800), "to" (4325), "of" (3917), "and" (3534), and "a" (2060). These words carry almost no meaningful information about the actual content of the novel. In contrast, the top 15 results from word_tokenization_normalization() using spaCy reveal words like "elizabeth" (639), "darcy" (428), "say" (426), "know" (394), and "bennet" (348) — words that are directly meaningful.
The difference comes from spaCy's preprocessing pipeline. It removes stop words (common words like "the" and "and"), filters out punctuation and numbers, discards tokens of two characters or fewer, and applies lemmatization — reducing words like "says" and "said" to their base form "say." This significantly cleans the results.
The spaCy approach gives far more meaningful output for content analysis. Based on the top 15 words, we can clearly infer that Pride and Prejudice centers on characters named Elizabeth, Darcy, Bennet, Bingley, Jane, and Mrs., and that social interaction — through words like "say," "know," "think," "come," and "see" — is central to the story. Removing filler words like "the" and "to" allows these meaningful themes and character names to surface clearly.
